File: src/server/DatabaseCommandsReceiver.java
package server;

import OrganizationDatabase.*;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import commands.server_side.*;
import lib.ExecutionStatus;
import network.client.DatabaseCommand;
import network.client.udp.ConnectionStatus;
import network.client.udp.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.SocketException;
import java.util.EnumMap;
import java.util.HashMap;
import java.util.Map;

public class DatabaseCommandsReceiver extends ServerWithConnectAndDisconnectCommand {
    private static final Logger logger = LogManager.getLogger(DatabaseCommandsReceiver.class);
    Map<User, OrganizationDatabase> usersDatabases = new HashMap<>();
    Map<DatabaseCommand, ServerSideCommand> commandMap = new EnumMap<>(DatabaseCommand.class);

    public DatabaseCommandsReceiver(int port) throws SocketException {
        super(port, "command");
        objectMapper.findAndRegisterModules();

        commandMap.put(DatabaseCommand.ADD, new AddCommand(this::statusOnlyCallback));
        commandMap.put(DatabaseCommand.ADD_IF_MAX, new AddIfMaxCommand(this::addIfMaxCommandCallback));
        commandMap.put(DatabaseCommand.SHOW, new ShowCommand(this::defaultDataSendCallback));
        commandMap.put(DatabaseCommand.CLEAR, new ClearCommand(this::statusOnlyCallback));
        commandMap.put(DatabaseCommand.INFO, new InfoCommand(this::defaultDataSendCallback));
        commandMap.put(DatabaseCommand.MAX_BY_FULL_NAME, new MaxByFullNameCommand(this::defaultDataSendCallback));
        commandMap.put(DatabaseCommand.REMOVE_HEAD, new RemoveHeadCommand(this::defaultDataSendCallback));
        commandMap.put(DatabaseCommand.REMOVE_BY_ID, new RemoveByIdCommand(this::statusOnlyCallback));
        commandMap.put(DatabaseCommand.SAVE, new SaveCommand(this::statusOnlyCallback));
        commandMap.put(DatabaseCommand.READ, new ReadCommand(this::statusOnlyCallback));
        commandMap.put(DatabaseCommand.REMOVE_ALL_BY_POSTAL_ADDRESS, new RemoveAllByPostalAddressCommand(this::statusOnlyCallback));
        commandMap.put(DatabaseCommand.UPDATE, new UpdateCommand(this::statusOnlyCallback));
    }

    private void addIfMaxCommandCallback(
            User user,
            OrganizationManagerInterface organizationManager,
            ExecutionStatus status,
            Exception error,
            Object... args
    ) throws IOException {
        assert args.length == 0;

        if (status == ExecutionStatus.FAILURE) {
            send(user, NetworkCode.NOT_A_MAXIMUM_ORGANIZATION, null);
        } else {
            send(user, NetworkCode.SUCCESS, null);
        }
    }

    private void defaultDataSendCallback(
            User user,
            OrganizationManagerInterface organizationManager,
            ExecutionStatus status,
            Exception error,
            Object... args
    ) throws IOException {
        assert (args.length == 1 && args[0] instanceof String) || status == ExecutionStatus.FAILURE;

        if (status == ExecutionStatus.FAILURE) {
            send(user, NetworkCode.FAILURE, null);
        } else {
            send(user, NetworkCode.SUCCESS, (String) args[0]);
        }
    }

    private void statusOnlyCallback(
            User user,
            OrganizationManagerInterface organizationManager,
            ExecutionStatus status,
            Exception error,
            Object... args
    ) throws IOException {
        assert args.length == 0;

        if (status == ExecutionStatus.FAILURE) {
            send(user, NetworkCode.FAILURE, null);
        } else {
            send(user, NetworkCode.SUCCESS, null);
        }
    }

    @Override
    protected void handlePacket(DatagramPacket packet) throws JsonProcessingException {
        ConnectionStatus status = handleConnectAndDisconnectCommands(packet);

        if (status != ConnectionStatus.CONNECTED) {
            return;
        }

        DatabaseCommand command = DatabaseCommand.valueOf(getCommandFromJson(packet));
        User user = getUserFromPacket(packet);

        logger.trace("Received command {}, from {}", command, user);

        commandMap.get(command).execute(
                getUserFromPacket(packet),
                usersDatabases.get(getUserFromPacket(packet))
        );

        Object[] arguments = switch (command) {
            case SHOW, INFO, CLEAR, MAX_BY_FULL_NAME, HISTORY, SUM_OF_ANNUAL_TURNOVER, REMOVE_HEAD -> new Object[]{};

            case ADD, ADD_IF_MAX, UPDATE -> new Object[]{getOrganization(getObjectNode(packet))};

            case REMOVE_BY_ID -> new Object[]{getInt(getObjectNode(packet))};

            case SAVE, READ -> new Object[]{getString(getObjectNode(packet))};

            case REMOVE_ALL_BY_POSTAL_ADDRESS -> new Object[]{getAddress(getObjectNode(packet))};

            default -> null; // can be join with the first case, but it might be used to treat null arguments as error
        };

        commandMap.get(command).execute(
                user,
                usersDatabases.get(getUserFromPacket(packet)),
                arguments
        );
    }

    @Override
    protected void handleConnectCommand(User user) {
        OrganizationDatabase org = new OrganizationDatabase();
        org.loadFromFile("/Users/vaskozlov/IdeaProjects/Programming5/test_database/test2.csv");
        usersDatabases.put(user, org);
    }

    @Override
    protected void handleDisconnectCommand(User user) {
        usersDatabases.remove(user);
    }

    private Organization getOrganization(JsonNode jsonNode) throws JsonProcessingException {
        return objectMapper.readValue(jsonNode.toString(), Organization.class);
    }

    private Address getAddress(JsonNode jsonNode) throws JsonProcessingException {
        return objectMapper.readValue(jsonNode.toString(), Address.class);
    }

    private String getString(JsonNode jsonNode) throws JsonProcessingException {
        return objectMapper.readValue(jsonNode.toString(), String.class);
    }

    private Integer getInt(JsonNode jsonNode) throws JsonProcessingException {
        return objectMapper.readValue(jsonNode.toString(), Integer.class);
    }

    protected JsonNode getObjectNode(DatagramPacket packet) throws JsonProcessingException {
        String result = new String(packet.getData(), 0, packet.getLength());
        JsonNode jsonNodeRoot = objectMapper.readTree(result);
        return jsonNodeRoot.get("object");
    }
}

File: src/server/DatabaseCommandsReceiver.kt
package server

import com.fasterxml.jackson.databind.JsonNode
import database.*
import lib.json.fromJson
import lib.net.udp.JsonHolder
import network.client.DatabaseCommand
import network.client.udp.User
import org.apache.logging.log4j.kotlin.Logging
import java.nio.file.Path
import kotlin.io.path.absolutePathString

class DatabaseCommandsReceiver(
    port: Int,
    userStoragePath: Path,
    private val databaseStoragePath: Path
) : ServerWithAuthorization(port, "command", AuthorizationManager(userStoragePath)), Logging {
    private var usersDatabases: MutableMap<AuthorizationHeader, OrganizationDatabase> = HashMap()
    private val commandArguments: MutableMap<DatabaseCommand, (AuthorizationHeader, JsonHolder) -> Any?> = mutableMapOf(
        DatabaseCommand.ADD to { _, jsonHolder ->
            jsonMapper.fromJson<Organization>(
                getObjectNode(
                    jsonHolder
                )
            )
        },
        DatabaseCommand.REMOVE_BY_ID to { _, jsonHolder ->
            jsonMapper.fromJson<Int>(
                getObjectNode(
                    jsonHolder
                )
            )
        },
        DatabaseCommand.REMOVE_ALL_BY_POSTAL_ADDRESS to { _, jsonHolder ->
            jsonMapper.fromJson<Address>(
                getObjectNode(
                    jsonHolder
                )
            )
        },
        DatabaseCommand.READ to { _, jsonHolder ->
            jsonMapper.fromJson<String>(
                getObjectNode(
                    jsonHolder
                )
            )
        },
        DatabaseCommand.SAVE to { authorizationHeader, _ ->
            getUserDatabaseFile(authorizationHeader).absolutePathString()
        },
        DatabaseCommand.EXIT to { _, _ -> null },
        DatabaseCommand.CLEAR to { _, _ -> null },
        DatabaseCommand.REMOVE_HEAD to { _, _ -> null },
        DatabaseCommand.MAX_BY_FULL_NAME to { _, _ -> null },
        DatabaseCommand.CLEAR to { _, _ -> null },
        DatabaseCommand.INFO to { _, _ -> null },
        DatabaseCommand.HISTORY to { _, _ -> null },
        DatabaseCommand.SUM_OF_ANNUAL_TURNOVER to { _, _ -> null },
    )

    init {
        commandArguments[DatabaseCommand.ADD_IF_MAX] = commandArguments[DatabaseCommand.ADD]!!
        commandArguments[DatabaseCommand.UPDATE] = commandArguments[DatabaseCommand.ADD]!!

        commandArguments[DatabaseCommand.SHOW] = commandArguments[DatabaseCommand.READ]!!

        val databaseDir = databaseStoragePath.toFile()
        databaseDir.mkdirs()
        require(databaseDir.isDirectory)
    }

    private suspend fun execute(
        command: DatabaseCommand,
        user: User,
        organizationManager: OrganizationManagerInterface,
        argument: Any?
    ): Result<Any?>? {
        return commandMap[command]?.execute(user, organizationManager, argument)
    }

    private fun getArgumentForTheCommand(
        command: DatabaseCommand,
        authorizationHeader: AuthorizationHeader,
        jsonHolder: JsonHolder
    ): Any? {
        return commandArguments[command]?.invoke(authorizationHeader, jsonHolder)
    }

    private suspend fun sendResult(
        user: User,
        result: Result<Any?>
    ) {
        val code = if (result.isSuccess) NetworkCode.SUCCESS else errorToNetworkCode(result.exceptionOrNull())
        logger.trace("Sending result to $user, code: $code")
        send(user, code, result.getOrNull())
    }

    override suspend fun handleAuthorized(
        user: User,
        authorizationHeader: AuthorizationHeader,
        jsonHolder: JsonHolder
    ) {
        val commandName = getCommandFromJson(jsonHolder)
        val database: OrganizationManagerInterface =
            usersDatabases.getOrPut(
                authorizationHeader
            ) { OrganizationDatabase(getUserDatabaseFile(authorizationHeader)) }

        if (!DatabaseCommand.entries.map { it.name }.contains(commandName)) {
            send(user, NetworkCode.NOT_SUPPOERTED_COMMAND, null)
            return
        }

        val command = DatabaseCommand.valueOf(commandName)
        logger.trace("Received command $command, from $user")

        if (!commandArguments.containsKey(command)) {
            logger.trace("Command: $command not found, from $user")
            send(user, NetworkCode.NOT_SUPPOERTED_COMMAND, null)
            return
        }

        logger.trace("Executing command: $command , from $user")
        val result =
            execute(command, user, database, getArgumentForTheCommand(command, authorizationHeader, jsonHolder))
        sendResult(user, result!!)
    }

    private fun getObjectNode(jsonHolder: JsonHolder): JsonNode {
        return jsonHolder.getNode("value")
    }

    private fun getUserDatabaseFile(authorizationHeader: AuthorizationHeader): Path {
        return databaseStoragePath.resolve("${authorizationHeader.login}.csv")
    }
}
